clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,176 @@
+//====- LoweringPrepareArm64CXXABI.cpp - Arm64 ABI specific code --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides ARM64 C++ ABI specific code that is used during LLVMIR
+// lowering prepare.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../IR/MissingFeatures.h"
+#include "LoweringPrepareItaniumCXXABI.h"
+
+#include <assert.h>
+
+using cir::LoweringPrepareCXXABI;
+using cir::MissingFeatures;
+
+using namespace mlir;
+using namespace mlir::cir;
+
+namespace {
+class LoweringPrepareAArch64CXXABI : public LoweringPrepareItaniumCXXABI {
+public:
+  LoweringPrepareAArch64CXXABI(clang::TargetCXXABI::Kind k) : Kind(k) {}
+  mlir::Value lowerVAArg(CIRBaseBuilderTy &builder,
+                         mlir::cir::VAArgOp op) override;
+
+private:
+  clang::TargetCXXABI::Kind Kind;
+  mlir::Value lowerGenericAArch64VAArg(CIRBaseBuilderTy &builder,
+                                       mlir::cir::VAArgOp op);
+};
+} // namespace
+
+LoweringPrepareCXXABI *
+LoweringPrepareCXXABI::createAArch64ABI(clang::TargetCXXABI::Kind k) {
+  return new LoweringPrepareAArch64CXXABI(k);
+}
+
+mlir::Value LoweringPrepareAArch64CXXABI::lowerGenericAArch64VAArg(
+    CIRBaseBuilderTy &builder, mlir::cir::VAArgOp op) {
+  auto loc = op->getLoc();
+  auto valist = op->getOperand(0);
+  auto opResTy = op.getType();
+  // front end should not produce non-scalar type of VAArgOp
+  bool isSupportedType =
+      opResTy.isa<mlir::cir::IntType, mlir::cir::SingleType,
+                  mlir::cir::PointerType, mlir::cir::BoolType,
+                  mlir::cir::DoubleType>();
+  assert(isSupportedType);
+
```

Comment:
When I compare this path with `AArch64.cpp`, I see a lack of ABI specific handling. One example:
```
ABIArgInfo AI = classifyArgumentType(Ty, /*IsVariadic=*/true,
                                       CGF.CurFnInfo->getCallingConvention());
  // Empty records are ignored for parameter passing purposes.
  if (AI.isIgnore()) {
    ...
  }

  bool IsIndirect = AI.isIndirect();

  llvm::Type *BaseTy = CGF.ConvertType(Ty);
  if (IsIndirect)
    BaseTy = llvm::PointerType::getUnqual(BaseTy);
  else if (AI.getCoerceToType())
    BaseTy = AI.getCoerceToType();
```
We need to also take those into account to make sure our lowering is accurate. CIRGen already has a `ABIArgInfo`, perhaps it should be moved to some common location usable by CIRGen and Lowering.

However, not sure how much is worth digging into that right now given @sitio-couto should be putting the necessary bits down for us to be able to classify arguments according to target information, etc. 

What I'd like to see for now are more functions added to `clang/lib/CIR/Dialect/IR/MissingFeatures.h` and more things like `assert(!MissingFeatures::AArch64ABI_IsIndirect());`, etc. Once a proper way to tag arguments show up we can come to this file and do the right thing.

Reply:
Another example: `IsHSA` or `isHomogeneousAggregate`, they kinda changes everything, but there are many other.

Reply:
When I looked at code and thought about details of ABI. There are two types of situations.
1. Situation excluded by assertion of supported type  assert(isSupportedType); this type includes isHomogeneousAggregate, as supported type has no aggregate type. And also things like alignment TyAlign.getQuantity() > 8 or isAggregateTypeForABI(Ty). No, types we are supporting now have no such concerns. Correct me if  I'm making wrong assumption about those supported types .

2. isInDirect, I'll do assert(!MissingFeatures::AArch64ABI_IsIndirect()) for now. However, I feel we might want to handle this early on when we generate cir.var_arg. I.E if argument is passed indirect, then cir.var_arg result type should reflect indirectness, so lowering algorithm doesn't have to worry about this ABI subtlety.

Reply:
Gotcha. Like we discussed offline, we should just make it more clear in code with comments/unimplemented-unrecheables so it's trackable when the ABI CC library lands.